### PR TITLE
Bug 2024212: Take a working copy of KUBECONFIG

### PR DESCRIPTION
**Pull request by derekhiggins**
Before each tests starts take a working copy of KUBECONFIG,
its will both protect the test from other tests that may
update the main KUBECONFIG file (by only readin it once
the chances of reading it at the wronge time are reduced).
Also it protects other tests from changes the currently
running test might make to KUBECONFIG.

This doesn't catch all use cases e.g. it misses CLI objects
created outside of a lt block but it seems to be enough to
make a noticable impact.

### CODE REVIEW

**Comment by derekhiggins** [comment]
Running these 4 tests in parallel is a good reproducer for the problem and fails about 80% of the time 

```
"[sig-cli] oc builds complex build webhooks CRUD [Skipped:Disconnected] [Suite:openshift/conformance/parallel]"
"[sig-instrumentation] Prometheus when installed on the cluster shouldn't have failing rules evaluation [Skipped:Disconnected] [Suite:openshift/conformance/parallel]"
"[sig-cli] oc adm storage-admin [Suite:openshift/conformance/parallel]"
"[sig-devex][Feature:Templates] templateinstance security tests  should pass security tests [Suite:openshift/conformance/parallel]"

```
This patch bring it down to about 15%, I appear to still be missing a failure path as the 15% that failed are also because of an empty KUBECONFIG

**Comment by derekhiggins** [comment]
/hold
Looks like some tests aren't happy with how I've done the cleanup of the tmp file, needs update

**Comment by andfasano** [comment]
IIUC, this means copying the kubeconfig file for each test test using the extended client (so, potentially, hundreds of them). As already pointed out, there are still a bunch of tests not using it.
So, given that the kubeconfig is a global resource, wouldn't be better to move all the tests that are touching it under the `serial` suite instead (and having a general test failure if someone touches it)?

**Comment by derekhiggins** [comment]
> IIUC, this means copying the kubeconfig file for each test test using the extended client (so, potentially, hundreds of them). 

Possibly more then hundreds, I've tested the copyToTmp code in a loop with 100,000 iterations and it took just over 10 seconds so the time penalty shouldn't be significant.

> As already pointed out, there are still a bunch of tests not using it. So, given that the kubeconfig is a global resource, wouldn't be better to move all the tests that are touching it under the `serial` suite instead (and having a general test failure if someone touches it)?

so it would fail if anything touches KUBECONFIG during the test run? I like that idea once we've fixed any jobs that cause it to fail there would be no need to move them to be serial?

**Comment by andfasano** [comment]
> > IIUC, this means copying the kubeconfig file for each test test using the extended client (so, potentially, hundreds of them).
> 
> Possibly more then hundreds, I've tested the copyToTmp code in a loop with 100,000 iterations and it took just over 10 seconds so the time penalty shouldn't be significant.
> 
> > As already pointed out, there are still a bunch of tests not using it. So, given that the kubeconfig is a global resource, wouldn't be better to move all the tests that are touching it under the `serial` suite instead (and having a general test failure if someone touches it)?
> 
> so it would fail if anything touches KUBECONFIG during the test run? I like that idea once we've fixed any jobs that cause it to fail there would be no need to move them to be serial?

If it's possible to fix them directly then I think so. The kubeconfig synthetic test initially could be something simple, ie verify at the end of the execution that the file wasn't touched.

**Comment by derekhiggins** [comment]
Maybe this is a good way to do it instead, it should also tell use which tests are trying to change the file
https://github.com/openshift/release/pull/23764

File: test/extended/util/client.go [hunk 1]
```diff
@@ -390,6 +393,21 @@ func (c *CLI) TeardownProject() {
...
+		c.adminConfigPath = c.origAdminConfigPath
+		c.origAdminConfigPath = ""
+	}
+}
+
+// SetupCLI takes a copy of KUBECONFIG to ensure the cli doesn't write to the original
+// and doesn't read the origional as often (reducing the likelyhood of being read while being written to by another test)
+func (c *CLI) SetupCLI() {
+	c.origAdminConfigPath = c.adminConfigPath
+	c.adminConfigPath = copyToTmp(c.adminConfigPath)
```
Comment (stbenjam) [inline test/extended/util/client.go:410]: Are we sure this is safe? In the parallel suite, are multiple tests using the same CLI?

Comment (derekhiggins) [reply test/extended/util/client.go:410]: I believe its safe, I've checked it with some tests in parallel and each one had its own CLI object. 
Also other methods in CLI  update c.configPath, if it was a problem the tests would have shown this up by now.

**Comment by andfasano** [comment]
> Maybe this is a good way to do it instead, it should also tell use which tests are trying to change the file [openshift/release#23764](https://github.com/openshift/release/pull/23764)

Yep, and we don't risk to miss those tests that are not using the go client

**Comment by stbenjam** [comment]
/lgtm
/hold

@deads2k Can you make sure this makes sense to you? I think the alternative I prefer in #23764 needs changes to ci-operator to allow specifying capabilities. For now, hopefully this PR makes metal ipi more stable.

**Pull request closed without merging**